source/extensions/transport_sockets/tls/context_impl.cc

```diff
@@ -567,61 +620,35 @@ std::vector<Ssl::PrivateKeyMethodProviderSharedPtr> ContextImpl::getPrivateKeyMe
   return providers;
 }
 
-bool ContextImpl::verifySubjectAltName(X509* cert,
-                                       const std::vector<std::string>& subject_alt_names) {
+bool ContextImpl::matchSubjectAltName(
+    X509* cert, const std::vector<Matchers::StringMatcherImpl>& match_subject_alt_name_list) {
   bssl::UniquePtr<GENERAL_NAMES> san_names(
       static_cast<GENERAL_NAMES*>(X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr)));
   if (san_names == nullptr) {
     return false;
   }
   for (const GENERAL_NAME* san : san_names.get()) {
-    switch (san->type) {
-    case GEN_DNS: {
-      ASN1_STRING* str = san->d.dNSName;
-      const char* dns_name = reinterpret_cast<const char*>(ASN1_STRING_data(str));
-      for (auto& config_san : subject_alt_names) {
-        if (dnsNameMatch(config_san, dns_name)) {
-          return true;
-        }
+    for (auto& config_san_matcher : match_subject_alt_name_list) {
+      if (config_san_matcher.match(generalNameAsString(san))) {
+        return true;
       }
-      break;
-    }
-    case GEN_URI: {
-      ASN1_STRING* str = san->d.uniformResourceIdentifier;
-      const char* uri = reinterpret_cast<const char*>(ASN1_STRING_data(str));
-      for (auto& config_san : subject_alt_names) {
-        if (config_san.compare(uri) == 0) {
-          return true;
-        }
-      }
-      break;
-    }
-    case GEN_IPADD: {
-      if (san->d.ip->length == 4) {
-        sockaddr_in sin;
-        sin.sin_port = 0;
-        sin.sin_family = AF_INET;
-        memcpy(&sin.sin_addr, san->d.ip->data, sizeof(sin.sin_addr));
-        Network::Address::Ipv4Instance addr(&sin);
-        for (auto& config_san : subject_alt_names) {
-          if (config_san == addr.ip()->addressAsString()) {
-            return true;
-          }
-        }
-      } else if (san->d.ip->length == 16) {
-        sockaddr_in6 sin6;
-        sin6.sin6_port = 0;
-        sin6.sin6_family = AF_INET6;
-        memcpy(&sin6.sin6_addr, san->d.ip->data, sizeof(sin6.sin6_addr));
-        Network::Address::Ipv6Instance addr(sin6);
-        for (auto& config_san : subject_alt_names) {
-          if (config_san == addr.ip()->addressAsString()) {
-            return true;
-          }
-        }
-      }
-      break;
     }
+  }
+  return false;
+}
+
+bool ContextImpl::verifySubjectAltName(X509* cert,
+                                       const std::vector<std::string>& subject_alt_names) {
+  bssl::UniquePtr<GENERAL_NAMES> san_names(
+      static_cast<GENERAL_NAMES*>(X509_get_ext_d2i(cert, NID_subject_alt_name, nullptr, nullptr)));
+  if (san_names == nullptr) {
+    return false;
+  }
+  for (const GENERAL_NAME* san : san_names.get()) {
+    for (auto& config_san : subject_alt_names) {
+      if (san->type == GEN_DNS ? dnsNameMatch(config_san, generalNameAsString(san).c_str())
+                               : config_san == generalNameAsString(san))
```

Comment:
make a temp for generalNameAsString(san), hoisting to the body of the outer-loop.